pkg/controller/certificaterequests/selfsigned/selfsigned_test.go

```diff
@@ -0,0 +1,356 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	clientcorev1 "k8s.io/client-go/listers/core/v1"
+
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1"
+	testpkg "github.com/jetstack/cert-manager/pkg/controller/test"
+	testfake "github.com/jetstack/cert-manager/pkg/controller/test/fake"
+	"github.com/jetstack/cert-manager/pkg/issuer"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+	"github.com/jetstack/cert-manager/test/unit/gen"
+)
+
+func generateCSR(t *testing.T, secretKey crypto.Signer, alg x509.SignatureAlgorithm) []byte {
+	asn1Subj, _ := asn1.Marshal(pkix.Name{
+		CommonName: "test",
+	}.ToRDNSequence())
+	template := x509.CertificateRequest{
+		RawSubject:         asn1Subj,
+		SignatureAlgorithm: alg,
+	}
+
+	csrBytes, err := x509.CreateCertificateRequest(rand.Reader, &template, secretKey)
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+
+	csr := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE REQUEST", Bytes: csrBytes})
+
+	return csr
+}
+
+func mustNoResponse(builder *testpkg.Builder, args ...interface{}) {
+	resp := args[0].(*issuer.IssueResponse)
+	if resp != nil {
+		builder.T.Errorf("unexpected response, exp='nil' got='%+v'", resp)
+	}
+}
+
+func TestSign(t *testing.T) {
+	skRSA, err := pki.GenerateRSAPrivateKey(2048)
+	if err != nil {
+		t.Errorf("failed to generate RSA private key: %s", err)
+		t.FailNow()
+	}
+
+	skEC, err := pki.GenerateECPrivateKey(256)
+	if err != nil {
+		t.Errorf("failed to generate ECDA private key: %s", err)
+		t.FailNow()
+	}
+
+	skAnotherRSA, err := pki.GenerateRSAPrivateKey(2048)
+	if err != nil {
+		t.Errorf("failed to generate RSA private key: %s", err)
+		t.FailNow()
+	}
+
+	csrBytes := generateCSR(t, skRSA, x509.SHA256WithRSA)
+	csrECBytes := generateCSR(t, skEC, x509.ECDSAWithSHA256)
+
+	skRSAPEMBytes := pki.EncodePKCS1PrivateKey(skRSA)
+	skAnotherRSAPEMBytes := pki.EncodePKCS1PrivateKey(skAnotherRSA)
+	skECPEMBytes, err := pki.EncodeECPrivateKey(skEC)
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+
+	rsaKeySecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-rsa-key",
+			Namespace: gen.DefaultTestNamespace,
+		},
+		Data: map[string][]byte{
+			corev1.TLSPrivateKeyKey: skRSAPEMBytes,
+		},
+	}
+	anotherRSAKeySecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-another-rsa-key",
+			Namespace: gen.DefaultTestNamespace,
+		},
+		Data: map[string][]byte{
+			corev1.TLSPrivateKeyKey: skAnotherRSAPEMBytes,
+		},
+	}
+	ecKeySecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-ecdsa-key",
+			Namespace: gen.DefaultTestNamespace,
+		},
+		Data: map[string][]byte{
+			corev1.TLSPrivateKeyKey: skECPEMBytes,
+		},
+	}
+	badKeySecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-bad-key",
+			Namespace: gen.DefaultTestNamespace,
+		},
+		Data: map[string][]byte{
+			corev1.TLSPrivateKeyKey: []byte("this is a bad key"),
+		},
+	}
```

Comment:
In future I think extracting the `cryptoBundle` structure out of the `certificate_request_test.go` file and utilising it here would be best, but we can leave that for later..